cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java

```diff
@@ -291,6 +310,19 @@ public void addOutput(String datasetName, Map<String, String> arguments) {
 
   @Override
   public void addOutput(String outputName, OutputFormatProvider outputFormatProvider) {
+    try {
```

Comment:
We're assuming here that sources/sinks that have the same name will have the same general properties. They can have runtime arguments that are different across runs. Hence it is okay to store the properties only once. 

That being said, today we don't have a way to way to distinguish properties from runtime arguments for a source or a sink. Both properties and arguments are combined together as configuration. Hence we'll end up storing properties + (runtime arguments of one source/sink) as the dataset properties.

I'll file a JIRA to address this in a later release.